### ISSUE TITLE
如何可视化每个类别的AP值？

### ISSUE THREAD

**Issue by Godk02** [lyuwenyu/RT-DETR#302]

**Comment by Godk02** [comment]
> `torch`版本的暂时不支持这功能

大佬后续会有这功能吗

**Comment by lyuwenyu** [comment]
1. 暂时没有开发的计划  ( 其实只需要去`coco eval`里按照你的改下应该就行 我记得之前`issue`里有人改成功了
2. 我不确定 `ultralytics` 这里的训练精度对齐了嘛，推理的话是转的`paddle`的参数

**Comment by Godk02** [comment]
> 1. 暂时没有开发的计划  ( 其实只需要去`coco eval`里按照你的改下应该就行 我记得之前`issue`里有人改成功了
> 2. 我不确定 `ultralytics` 这里的训练精度对齐了嘛，推理的话是转的`paddle`的参数

非常感谢您的回复！！

**Comment by DaCheng1823** [comment]
能详细说一下测速使用的部分代码和工具怎么使用的吗？没太看懂怎么使用的

**Comment by Godk02** [comment]
> 能详细说一下测速使用的部分代码和工具怎么使用的吗？没太看懂怎么使用的

issue#53

**Comment by lyuwenyu** [comment]
推理参考 https://github.com/lyuwenyu/RT-DETR/issues/95

**Comment by lyuwenyu** [comment]
> > `torch`版本的暂时不支持这功能
> 
> 大佬后续会有这功能吗

@Godk02  后续`rtdetrv2`里会支持

**Comment by Godk02** [comment]
**`大佬，我想请教一下，改了coco_eval.py中的summarize函数**
`def summarize(self):
        for iou_type, coco_eval in self.coco_eval.items():
           print("IoU metric: {}".format(iou_type))
            coco_eval.summarize()`
 **改为了**
 `def summarize(self):
        for iou_type, coco_eval in self.coco_eval.items():
            print("IoU metric: {}".format(iou_type))
            coco_eval.summarize()

            # Print AP for each category at IoU=0.50
            if iou_type == "bbox":
                aps = []
                for cat_id in range(7):  # Assuming class IDs range from 1 to 7
                    cat_name = coco_eval.cocoGt.cats[cat_id + 1]['name']  # Adjusted cat_id
                    if coco_eval.eval['precision'].shape[4] > 0:  # Check if index 0 is within bounds
                        precision = coco_eval.eval['precision'][0, :, cat_id, 0, 1]  # AP at IoU=0.50
                        if not np.isnan(np.mean(precision)):  # Check if precision is not NaN
                            ap = np.mean(precision)
                            aps.append(ap)
                            print("Category: {}, AP@0.50: {:.3f}".format(cat_name, ap))
                        else:
                            print("Category: {}, No detections for AP@0.50".format(cat_name))
                    else:
                        print("Category: {}, No AP@0.50 value available".format(cat_name))
                if aps:  # Check if list is not empty
                    print("Mean AP@0.50 across all categories: {:.3f}".format(np.mean(aps)))
                else:
                    print("Mean AP@0.50 across all categories: No detections")`
 **输出结果如下:**
 WARNING:torch.distributed.run:
*****************************************
Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
*****************************************
Initialized distributed mode...
loading annotations into memory...
Done (t=0.04s)
creating index...
index created!
resume from /scratch/chenjun3/qiankun/RT-DETR-main/rtdetr_pytorch/logs/checkpoint0124.pth
Loading last_epoch
Loading model.state_dict
Loading ema.state_dict
Test:  [ 0/32]  eta: 0:02:27    time: 4.6004  data: 2.3144  max mem: 2399
Test:  [10/32]  eta: 0:00:18    time: 0.8587  data: 0.2344  max mem: 2399
Test:  [20/32]  eta: 0:00:08    time: 0.4997  data: 0.0289  max mem: 2399
Test:  [30/32]  eta: 0:00:01    time: 0.4738  data: 0.0308  max mem: 2399
Test:  [31/32]  eta: 0:00:00    time: 0.4453  data: 0.0297  max mem: 2399
Test: Total time: 0:00:19 (0.5989 s / it)
Averaged stats: 
Accumulating evaluation results...
DONE (t=2.62s).
IoU metric: bbox
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.325
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.666
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.279
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.047
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.205
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.349
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.250
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.388
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.425
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.068
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.302
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.418
Category: x, AP@0.50: 0.608
Category: zh, AP@0.50: 0.450
Category: zhu, AP@0.50: 0.503
Category: die, AP@0.50: 0.621
Category: zssb, AP@0.50: 0.658
Category: zstb, AP@0.50: 0.782
Category: d, AP@0.50: 0.956
Mean AP@0.50 across all categories: 0.654
 我这里的Mean AP@0.50 across all categories: 0.654与Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.666不一致，计算方法不同，实验表格可以按照我改的填不

**Comment by Godk02** [comment]
**大佬，我想请教一下，改了coco_eval.py中的summarize函数** 
`def summarize(self):
        for iou_type, coco_eval in self.coco_eval.items():
            print("IoU metric: {}".format(iou_type))
            coco_eval.summarize()

            # Print AP for each category at IoU=0.50
            if iou_type == "bbox":
                aps = []
                for cat_id in range(7):  # Assuming class IDs range from 1 to 7
                    cat_name = coco_eval.cocoGt.cats[cat_id + 1]['name']  # Adjusted cat_id
                    if coco_eval.eval['precision'].shape[4] > 0:  # Check if index 0 is within bounds
                        precision = coco_eval.eval['precision'][0, :, cat_id, 0, 1]  # AP at IoU=0.50
                        if not np.isnan(np.mean(precision)):  # Check if precision is not NaN
                            ap = np.mean(precision)
                            aps.append(ap)
                            print("Category: {}, AP@0.50: {:.3f}".format(cat_name, ap))
                        else:
                            print("Category: {}, No detections for AP@0.50".format(cat_name))
                    else:
                        print("Category: {}, No AP@0.50 value available".format(cat_name))
                if aps:  # Check if list is not empty
                    print("Mean AP@0.50 across all categories: {:.3f}".format(np.mean(aps)))
                else:
                    print("Mean AP@0.50 across all categories: No detections")`
**输出结果如下:**
WARNING:torch.distributed.run:
Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed.
Initialized distributed mode...
loading annotations into memory...
Done (t=0.04s)
creating index...
index created!
resume from /scratch/chenjun3/qiankun/RT-DETR-main/rtdetr_pytorch/logs/checkpoint0124.pth
Loading last_epoch
Loading model.state_dict
Loading ema.state_dict
Test: [ 0/32] eta: 0:02:27 time: 4.6004 data: 2.3144 max mem: 2399
Test: [10/32] eta: 0:00:18 time: 0.8587 data: 0.2344 max mem: 2399
Test: [20/32] eta: 0:00:08 time: 0.4997 data: 0.0289 max mem: 2399
Test: [30/32] eta: 0:00:01 time: 0.4738 data: 0.0308 max mem: 2399
Test: [31/32] eta: 0:00:00 time: 0.4453 data: 0.0297 max mem: 2399
Test: Total time: 0:00:19 (0.5989 s / it)
Averaged stats:
Accumulating evaluation results...
DONE (t=2.62s).
IoU metric: bbox
Average Precision (AP) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.325
Average Precision (AP) @[ IoU=0.50 | area= all | maxDets=100 ] = 0.666
Average Precision (AP) @[ IoU=0.75 | area= all | maxDets=100 ] = 0.279
Average Precision (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.047
Average Precision (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.205
Average Precision (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.349
Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 1 ] = 0.250
Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 10 ] = 0.388
Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.425
Average Recall (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.068
Average Recall (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.302
Average Recall (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.418
Category: x, AP@0.50: 0.608
Category: zh, AP@0.50: 0.450
Category: zhu, AP@0.50: 0.503
Category: die, AP@0.50: 0.621
Category: zssb, AP@0.50: 0.658
Category: zstb, AP@0.50: 0.782
Category: d, AP@0.50: 0.956
Mean AP@0.50 across all categories: 0.654
**我这里的Mean AP@0.50 across all categories: 0.654与Average Precision (AP) @[ IoU=0.50 | area= all | maxDets=100 ] = 0.666不一致，计算方法不同，实验表格可以按照我改的填不**

**Comment by Godk02** [comment]
> > 1. 暂时没有开发的计划  ( 其实只需要去`coco eval`里按照你的改下应该就行 我记得之前`issue`里有人改成功了
> > 2. 我不确定 `ultralytics` 这里的训练精度对齐了嘛，推理的话是转的`paddle`的参数
> 
> 非常感谢您的回复！！

**大佬我将coco_eval.py中的summarize函数做了修改，如下：**
`    def summarize(self):
        for iou_type, coco_eval in self.coco_eval.items():
            print("IoU metric: {}".format(iou_type))
            coco_eval.summarize()

            # Print AP for each category at IoU=0.50
            if iou_type == "bbox":
                aps = []
                for cat_id in range(7):  # Assuming class IDs range from 1 to 7
                    cat_name = coco_eval.cocoGt.cats[cat_id + 1]['name']  # Adjusted cat_id
                    if coco_eval.eval['precision'].shape[4] > 0:  # Check if index 0 is within bounds
                        precision = coco_eval.eval['precision'][0, :, cat_id, 0, 1]  # AP at IoU=0.50
                        if not np.isnan(np.mean(precision)):  # Check if precision is not NaN
                            ap = np.mean(precision)
                            aps.append(ap)
                            print("Category: {}, AP@0.50: {:.3f}".format(cat_name, ap))
                        else:
                            print("Category: {}, No detections for AP@0.50".format(cat_name))
                    else:
                        print("Category: {}, No AP@0.50 value available".format(cat_name))
                if aps:  # Check if list is not empty
                    print("Mean AP@0.50 across all categories: {:.3f}".format(np.mean(aps)))
                else:
                    print("Mean AP@0.50 across all categories: No detections")`
 **输出结果如下：**
 `WARNING:torch.distributed.run:
*****************************************
Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
*****************************************
Initialized distributed mode...
loading annotations into memory...
Done (t=0.04s)
creating index...
index created!
resume from /scratch/chenjun3/qiankun/RT-DETR-main/rtdetr_pytorch/logs/checkpoint0124.pth
Loading last_epoch
Loading model.state_dict
Loading ema.state_dict
Test:  [ 0/32]  eta: 0:02:27    time: 4.6004  data: 2.3144  max mem: 2399
Test:  [10/32]  eta: 0:00:18    time: 0.8587  data: 0.2344  max mem: 2399
Test:  [20/32]  eta: 0:00:08    time: 0.4997  data: 0.0289  max mem: 2399
Test:  [30/32]  eta: 0:00:01    time: 0.4738  data: 0.0308  max mem: 2399
Test:  [31/32]  eta: 0:00:00    time: 0.4453  data: 0.0297  max mem: 2399
Test: Total time: 0:00:19 (0.5989 s / it)
Averaged stats: 
Accumulating evaluation results...
DONE (t=2.62s).
IoU metric: bbox
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.325
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.666
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.279
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.047
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.205
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.349
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.250
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.388
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.425
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.068
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.302
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.418
Category: xuansh(AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.388
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.425
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.068
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.302
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.418
Category: x, AP@0.50: 0.608
Category: z, AP@0.50: 0.450
Category: zh, AP@0.50: 0.503
Category: die, AP@0.50: 0.621
Category: zssb, AP@0.50: 0.658
Category: zstb, AP@0.50: 0.782
Category: d, AP@0.50: 0.956
Mean AP@0.50 across all categories: 0.654
`
**请问实验表格可以按照我改的填写吗，因为我注意到mAP的计算方法不太一样**

**Comment by xyb1314** [comment]
> > 1. 暂时没有开发的计划  ( 其实只需要去`coco eval`里按照你的改下应该就行 我记得之前`issue`里有人改成功了
> > 2. 我不确定 `ultralytics` 这里的训练精度对齐了嘛，推理的话是转的`paddle`的参数
> 
> 非常感谢您的回复！！

大佬,你知道怎么改吗？显示每个类别的mAP

**Comment by Godk02** [comment]
> > > 1. 暂时没有开发的计划  ( 其实只需要去`coco eval`里按照你的改下应该就行 我记得之前`issue`里有人改成功了
> > > 2. 我不确定 `ultralytics` 这里的训练精度对齐了嘛，推理的话是转的`paddle`的参数
> > 
> > 
> > 非常感谢您的回复！！
> 
> 大佬,你知道怎么改吗？显示每个类别的mAP
![image](https://github.com/user-attachments/assets/3d2c2f17-bda5-49f7-8610-5fc6dcc0a103)

可以参考我发布的https://github.com/Godk02/RTDETR_val_AP

**Comment by xinyuzhang1026** [comment]
> > > > 1. 暂时没有开发的计划（只要去`coco eval`里按照你的改下就应该行我记得之前`issue`有人改成功了
> > > > 2. 我不确定`ultralytics`这里的训练精度，推理问题是转的`paddle`参数
> > > 
> > > 
> > > 非常感谢您的回复！！
> > 
> > 
> > 大佬，你知道怎么改吗？显示每个类别的地图
> 
> 可以参考我发布的https://github.com/Godk02/RTDETR_val_AP

大佬们，我替换了这个coco eval，train的时候运行不了，是要训练完之后，用best.pth预测的时候才替换掉coco eval，预测的时候出现每一类的AP吗？该输入什么命令进行预测？向大佬们求助！

**Comment by Godk02** [comment]
> > > > > 1. 暂时没有开发的计划（只要去`coco eval`里按照你的改下就应该行我记得之前`issue`有人改成功了
> > > > > 2. 我不确定`ultralytics`这里的训练精度，推理问题是转的`paddle`参数
> > > > 
> > > > 
> > > > 非常感谢您的回复！！
> > > 
> > > 
> > > 大佬，你知道怎么改吗？显示每个类别的地图
> > 
> > 
> > 可以参考我发布的https://github.com/Godk02/RTDETR_val_AP
> 
> 大佬们，我替换了这个coco eval，train的时候运行不了，是要训练完之后，用best.pth预测的时候才替换掉coco eval，预测的时候出现每一类的AP吗？该输入什么命令进行预测？向大佬们求助！

报的啥错发一下

**Comment by Godk02** [comment]
> > > > > 1. 暂时没有开发的计划（只要去`coco eval`里按照你的改下就应该行我记得之前`issue`有人改成功了
> > > > > 2. 我不确定`ultralytics`这里的训练精度，推理问题是转的`paddle`参数
> > > > 
> > > > 
> > > > 非常感谢您的回复！！
> > > 
> > > 
> > > 大佬，你知道怎么改吗？显示每个类别的地图
> > 
> > 
> > 可以参考我发布的https://github.com/Godk02/RTDETR_val_AP
> 
> 大佬们，我替换了这个coco eval，train的时候运行不了，是要训练完之后，用best.pth预测的时候才替换掉coco eval，预测的时候出现每一类的AP吗？该输入什么命令进行预测？向大佬们求助！

有eval的代码啊，train的指令加个test only就行了
